src/auto-evo/simulation/food_source/CompoundFoodSource.cs

```diff
@@ -0,0 +1,61 @@
+﻿public class CompoundFoodSource : IFoodSource
+{
+    private readonly Compound glucose = SimulationParameters.Instance.GetCompound("glucose");
+    private readonly Compound atp = SimulationParameters.Instance.GetCompound("atp");
+
+    private BiomeConditions biomeConditions;
+    private Compound compound;
+    private float totalCompound;
+
+    public CompoundFoodSource(Patch patch, Compound compound)
+    {
+        biomeConditions = patch.Biome;
+        this.compound = compound;
+        if (patch.Biome.Compounds.ContainsKey(compound))
+        {
+            totalCompound = patch.Biome.Compounds[compound].Density * patch.Biome.Compounds[compound].Amount;
+        }
+        else
+        {
+            totalCompound = 0.0f;
+        }
+    }
+
+    public float FitnessScore(Species species)
+    {
+        var microbeSpecies = (MicrobeSpecies)species;
+
+        var compoundUseScore = 0.0f;
+        foreach (var organelle in microbeSpecies.Organelles)
+        {
+            foreach (var process in organelle.Definition.RunnableProcesses)
+            {
+                if (process.Process.Inputs.ContainsKey(compound))
+                {
+                    if (process.Process.Outputs.ContainsKey(glucose))
+                    {
+                        compoundUseScore += process.Process.Outputs[glucose]
+                            / process.Process.Inputs[compound] / Constants.AUTO_EVO_GLUCOSE_USE_SCORE_DIVISOR;
+                    }
+
+                    if (process.Process.Outputs.ContainsKey(atp))
+                    {
+                        compoundUseScore += process.Process.Outputs[atp]
+                            / process.Process.Inputs[compound] / Constants.AUTO_EVO_ATP_USE_SCORE_DIVISOR;
+                    }
+                }
+            }
+        }
+
+        var energyCost = ProcessSystem.ComputeEnergyBalance(
+            microbeSpecies.Organelles.Organelles,
+            biomeConditions, microbeSpecies.MembraneType).FinalBalanceStationary;
```

Comment:
You have the very same code in EnvirontalFoodSource, minus ATP (which could be added back in Environmental food source as the class is now more generic, btw). You might want to factor it somehow. 

Perhaps an intermediate interface for these two classes, helpers, or even rely on the `MicrobeSpecies.cs` class for some parts?

(EDIT: After further reading it seems like `MicrobeInternalCalculation` could be of use, although it seems linked to microbes rather than species)

Reply:
This gets into a more esoteric argument about software design (which I'm happy to get in if you really want to), but I would rather let this stand.  In short, code is duplicated in the literal sense, but these two algorithms are responsible for different things, and I expect them to evolve in separate directions over time.

Reply:
I agree but for this part.
> I expect them to evolve in separate directions over time
Still, I can't affirm that I am right in my expectations so I do not have further objection to keeping them duplicated.

Nonetheless, I'd think that compound use score could be useful in general and could thus benefit from a perhaps parametric computation (e.g. passing multiplier by output compound). This is really at the edge of the PR's scope, so I'm open to leaving it for a potential future refactory.